compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp

```diff
@@ -362,7 +261,319 @@ static FailureOr<Operation *> tryRaiseToView(linalg::GenericOp linalgOp,
+  }
+  return std::nullopt;
+}
+
+class FuseMatmulTranspose : public OpRewritePattern<linalg::MatmulOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(matmulOp)) {
+      return failure();
+    }
+
+    std::optional<Value> newRhs = matchATransposeBMatmul(matmulOp);
+    if (!newRhs) {
+      return rewriter.notifyMatchFailure(matmulOp,
+                                         "Failed to match transpose on RHS");
+    }
+
+    Value lhs = matmulOp.getDpsInputOperand(0)->get();
+    Value init = matmulOp.getDpsInitOperand(0)->get();
+    rewriter.setInsertionPoint(matmulOp);
+    SmallVector<NamedAttribute> attrs = getPrunedAttributeList(matmulOp);
+    rewriter.replaceOpWithNewOp<linalg::MatmulTransposeBOp>(
+        matmulOp, ValueRange{lhs, *newRhs}, ValueRange{init}, attrs);
+    return success();
+  }
+};
+
+class FuseBatchMatmulTranspose
+    : public OpRewritePattern<linalg::BatchMatmulOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::BatchMatmulOp bmmOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(bmmOp)) {
+      return failure();
+    }
+
+    std::optional<Value> newRhs = matchATransposeBBatchMatmul(bmmOp);
+    if (!newRhs) {
+      return rewriter.notifyMatchFailure(bmmOp,
+                                         "Failed to match transpose on RHS");
+    }
+
+    Value lhs = bmmOp.getDpsInputOperand(0)->get();
+    Value init = bmmOp.getDpsInitOperand(0)->get();
+    rewriter.setInsertionPoint(bmmOp);
+    SmallVector<NamedAttribute> attrs = getPrunedAttributeList(bmmOp);
+    rewriter.replaceOpWithNewOp<linalg::BatchMatmulTransposeBOp>(
+        bmmOp, ValueRange{lhs, *newRhs}, ValueRange{init}, attrs);
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Fill Raising
+//===----------------------------------------------------------------------===//
+
+// Method to match a linalg.generic op representing a linalg.fill op. Returns
+// the fill value (input operand to linalg.fill) on success.
+std::optional<Value> matchGenericFill(linalg::LinalgOp linalgOp) {
+  if (isa<linalg::GenericOp>(linalgOp.getOperation()) &&
+      linalgOp.getNumDpsInputs() == 0 && linalgOp.getNumDpsInits() == 1 &&
+      linalgOp.getNumParallelLoops() == linalgOp.getNumLoops() &&
+      linalgOp.getIndexingMapsArray()[0].isIdentity()) {
+    // Check that the op body is only a linalg.yield op.
+    Value yieldOperand;
+    for (Operation &bodyOp : linalgOp.getBlock()->getOperations()) {
+      if (isa<linalg::YieldOp>(bodyOp)) {
+        yieldOperand = bodyOp.getOperand(0);
+      } else {
+        return std::nullopt;
+      }
+    }
```

Comment:
All the structured linalg ops must have linalg.yield as a terminator, so we can make this check much simpler. I.e., we can check if the body has only one element.

https://github.com/llvm/llvm-project/blob/a0c6b8aef853eedaa0980f07c0a502a5a8a9740e/mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td#L24-L34